### ISSUE TITLE
Issue with SP clustering

### ISSUE THREAD

**Issue by arrk-shoba** [toyokazu/omniauth-shibboleth#13]
Hi,

We are supporting a multi-tenant application. Most of the time, there is substantial load on our application. Our technology stack involves:
Rails 3.2.13
Ruby 1.9.3
Running on Apache/2.2.15 (Unix) with Passenger(3.0.11).
Related Gems:
devise (2.2.7)
oauth (0.4.7)
oauth2 (0.8.1)
omniauth (1.1.4)
omniauth-oauth (1.0.1)
omniauth-oauth2 (1.0.3)
omniauth-shibboleth (1.0.8)
We have configured shibboleth2 configuration to use the StorageService type="Memory"
We have 2-3 application servers running at any point in time. 
Off late, many of our clients have complained that they are unable to login via Shibboleth SSO. On debugging, what we observed is that, the issue is more frequently occurring on Chrome browser on Mac OS.
In the Browser Network window, we see that the user is not treated as logged in, even after successful authentication at IDP. If the user clicks the 'Sign in' link (which redirects to <domain.com?/users/auth/shibboleth), the system treats the user as logged in. This is very confusing for our clients.
Can you please help us understand where the issue might be? Is it at omniauth-shibboleth, devise or browser cookie's erratic behaviour? We are unable to conclude this issue.

Also, would you suggest to move the StorageService type="ODBC" instead of Memory since the Application servers (hence the SPs) are in auto-scale mode? If yes, I would need some documentation/tutorial for configuration of the ODBC StorageService.
When I gave it a try, the callback URL is not initiated at all - I am still wondering why has the omniauth_controller's callback provider method is not invoked at all.

regards,
Shoba

**Comment by toyokazu** [comment]
Hi,

In the title, you mention 'clustering'. Does it mean that several Apache servers are running and clustered by Apache itself? Or it just means 'multi-tenanted'? Please let me make sure that.

 In order to find where is the root cause, you can use :debug mode of omniauth-shibboleth.
https://github.com/toyokazu/omniauth-shibboleth#debug-mode

You can check the parameters like Shib-Session-ID are properly passed to the Rack and Rails. If the parameters are not passed to the Rack, the root cause resides in Shibboleth SP (mod_shib or shibd). If they are properly passed, it resides in Rack or Rails (including omniauth-shibboleth).

Would you try to check that?

Best Regards,

**Comment by arrk-shoba** [comment]
Hello Toyokazu San,

Several Apache servers are running. They are not clustered by Apache or anything. The servers are hosted in AWS and managed via Scalr. The application is multi-tenant system. 

I shall try your suggestion and update the ticket.

regards,
Shoba

**Comment by arrk-shoba** [comment]
Hello Toyokazu San,

The issue is resolved.
I followed following steps:

1.	Since we are using Postgres 9.3, install postgres ODBC libs:
Steps to install postgres library & ODBC

rpm -ivh http://yum.postgresql.org/9.3/redhat/rhel-5.4-x86_64/pgdg-centos93-9.3-1.noarch.rpm

yum install postgresql93-odbc*

yum install -y postgresql-libs

ln -s /usr/lib64/libpq.so.5 /usr/lib64/libpq.so
ln -s /usr/pgsql-9.3/lib/psqlodbc.so /usr/lib64/psqlodbc.so

2.	Created database and tables:
CREATE TABLE version (
    major int NOT NULL,
    minor int NOT NULL
    );
INSERT INTO version VALUES (1,0);
  
CREATE TABLE strings (
    context varchar(255) NOT NULL,
    id      varchar(255) NOT NULL,
    expires timestamp    NOT NULL,
    version smallint     NOT NULL,
    value   varchar(255) NOT NULL,
    PRIMARY KEY (context, id)
    );
  
CREATE TABLE texts (
    context varchar(255) NOT NULL,
    id      varchar(255) NOT NULL,
    expires timestamp    NOT NULL,
    version smallint     NOT NULL,
    value   text         NOT NULL,
    PRIMARY KEY (context, id)
);

3.	Modified shibboleth2.xml:

<OutOfProcess logger="shibd.logger">
<Extensions>
  <Library path="odbc-store.so" fatal="true"/>
</Extensions>
</OutOfProcess>

<StorageService type="ODBC" id="db" cleanupInterval="900">
        <ConnectionString>        DRIVER=PostgreSQL;SERVER=database_host;PORT=5432;Uid=db_user;PASSWORD=db_pwd;DATABASE=db_name
        </ConnectionString>
    </StorageService>
    <SessionCache type="StorageService" StorageService="db" cacheAssertions="false"
                  cacheTimeout="3600" inprocTimeout="900" cleanupInterval="900"/>
    <ReplayCache StorageService="db"/>
    <ArtifactMap StorageService="db" artifactTTL="180"/>

<SessionInitiator type="Chaining" Location="/Login" isDefault="true" id="Default"
        relayState="ss:db" forceAuthn="true" entityID="">
    <SessionInitiator type="SAML2" acsIndex="1" template="bindingTemplate.html"/>
</SessionInitiator>
 
<LogoutInitiator type="Chaining" Location="/Logout" relayState="ss:db">
    <LogoutInitiator type="SAML2" template="bindingTemplate.html"/>
    <LogoutInitiator type="Local"/>
</LogoutInitiator>


4.	Restart Apache and Shibd services.

You may close this issue.

regards,
Shoba

**Comment by arrk-shoba** [comment]
Hello Toyokazu San,

The issue is resolved.
I followed following steps:

1.	Since we are using Postgres 9.3, install postgres ODBC libs:
Steps to install postgres library & ODBC

rpm -ivh http://yum.postgresql.org/9.3/redhat/rhel-5.4-x86_64/pgdg-centos93-9.3-1.noarch.rpm

yum install postgresql93-odbc*

yum install -y postgresql-libs

ln -s /usr/lib64/libpq.so.5 /usr/lib64/libpq.so
ln -s /usr/pgsql-9.3/lib/psqlodbc.so /usr/lib64/psqlodbc.so

2.	Created database and tables:
CREATE TABLE version (
    major int NOT NULL,
    minor int NOT NULL
    );
INSERT INTO version VALUES (1,0);
  
CREATE TABLE strings (
    context varchar(255) NOT NULL,
    id      varchar(255) NOT NULL,
    expires timestamp    NOT NULL,
    version smallint     NOT NULL,
    value   varchar(255) NOT NULL,
    PRIMARY KEY (context, id)
    );
  
CREATE TABLE texts (
    context varchar(255) NOT NULL,
    id      varchar(255) NOT NULL,
    expires timestamp    NOT NULL,
    version smallint     NOT NULL,
    value   text         NOT NULL,
    PRIMARY KEY (context, id)
);

3.	Modified shibboleth2.xml:

<OutOfProcess logger="shibd.logger">
<Extensions>
  <Library path="odbc-store.so" fatal="true"/>
</Extensions>
</OutOfProcess>

<StorageService type="ODBC" id="db" cleanupInterval="900">
        <ConnectionString>        DRIVER=PostgreSQL;SERVER=database_host;PORT=5432;Uid=db_user;PASSWORD=db_pwd;DATABASE=db_name
        </ConnectionString>
    </StorageService>
    <SessionCache type="StorageService" StorageService="db" cacheAssertions="false"
                  cacheTimeout="3600" inprocTimeout="900" cleanupInterval="900"/>
    <ReplayCache StorageService="db"/>
    <ArtifactMap StorageService="db" artifactTTL="180"/>

And following in the ApplicationDefault's Session or ApplicationOverride's Session tag:
<SessionInitiator type="Chaining" Location="/Login" isDefault="true" id="Default"
        relayState="ss:db" forceAuthn="true" entityID="">
    <SessionInitiator type="SAML2" acsIndex="1" template="bindingTemplate.html"/>
</SessionInitiator>
 
<LogoutInitiator type="Chaining" Location="/Logout" relayState="ss:db">
    <LogoutInitiator type="SAML2" template="bindingTemplate.html"/>
    <LogoutInitiator type="Local"/>
</LogoutInitiator>


4.	Restart Apache and Shibd services.

You may close this issue.

regards,
Shoba

**Comment by arrk-shoba** [comment]
Hello Toyokazu San,

The issue is resolved.
I followed following steps:

1.	Since we are using Postgres 9.3, install postgres ODBC libs:
Steps to install postgres library & ODBC

rpm -ivh http://yum.postgresql.org/9.3/redhat/rhel-5.4-x86_64/pgdg-centos93-9.3-1.noarch.rpm

yum install postgresql93-odbc*

yum install -y postgresql-libs

ln -s /usr/lib64/libpq.so.5 /usr/lib64/libpq.so
ln -s /usr/pgsql-9.3/lib/psqlodbc.so /usr/lib64/psqlodbc.so

Then follow the steps on https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPODBCStorageService

regards,
Shoba

**Comment by toyokazu** [comment]
Thank you for your report.
Recording the route cause may help the other people to solve the problems :)

Best

**Comment by arrk-shoba** [comment]
Hello Toyokazu San,

When I tried using ODBC StorageService the first time, callback was not initiated. I tried multiple things like restarting apache, shibd services. The Shibboleth session was generated; however control never reached callback. Due to this, I had posted the issue here to get your insight on the same.

After your reply regarding debug mode, I resumed my older configuration (using Memory StorageService). Checked in debug mode and it was working. I again modified the configuration to use the ODBC one. Created the tables for storing shibd session afresh. Restarted apache and shibd. This time, in the first go itself, I could see the callback being initiated.

Hence I do not have the exact root cause for the issue. One thing I learnt is, if we have a load balanced app servers, then ODBC storage service is more reliable than Memory one.

regards,
Shoba
